presto-main/src/main/java/com/facebook/presto/tz/JdkBasedDateTimeZone.java

```diff
@@ -0,0 +1,114 @@
+        extends DateTimeZone
+{
+    private final ZoneRules zoneRules;
+
+    public JdkBasedDateTimeZone(String zoneId)
+    {
+        super(zoneId);
+        this.zoneRules = ZoneId.of(zoneId).getRules();
+    }
+
+    @Override
+    public String getNameKey(long instant)
+    {
+        // It is ok to return a dummy value here.
+        // getNameKey is used in 6 instances in Joda
+        // * `DateTimeZone.getShortName`. Assuming DefaultNameProvider is being used,
+        //   it then invokes `DefaultNameProvider.getShortName`, which in turn invokes
+        //   `DefaultNameProvider.getNameSet`, which ignores the passed in nameKey,
+        //   other than validating that it's not null.
+        // * `DateTimeZone.getName`. Same as above.
+        // * `CachedDateTimeZone.Info.getNameKey`. It is used to implement getNameKey.
+        // * `DateTimeZoneBuilder.PrecalculatedZone.getNameKey`. Same as above.
+        // * `DateTimeZoneBuilder.writeTo`. It only uses getNameKey when the object
+        //   is an instance of `FixedDateTimeZone`.
+        // * `ZoneInfoCompiler.test`. It's only invoked with built-in implementations
+        //   of `DateTimeZone`.
+
+        // Use an obviously incorrect and easily searchable value here
+        return "presto-name-key-not-provided";
+    }
+
+    @Override
+    public int getOffset(long instant)
+    {
+        return zoneRules.getOffset(Instant.ofEpochMilli(instant)).getTotalSeconds() * 1000;
+    }
+
+    @Override
+    public int getStandardOffset(long instant)
+    {
+        return zoneRules.getStandardOffset(Instant.ofEpochMilli(instant)).getTotalSeconds() * 1000;
+    }
+
+    @Override
+    public boolean isFixed()
+    {
+        return zoneRules.isFixedOffset();
+    }
+
+    @Override
+    public long nextTransition(long instant)
+    {
+        ZoneOffsetTransition nextTransition = zoneRules.nextTransition(Instant.ofEpochMilli(instant));
+        if (nextTransition == null) {
+            // this is after the last transition
+            return instant;
+        }
+        return nextTransition.toEpochSecond() * 1000;
+    }
+
+    @Override
+    public long previousTransition(long instant)
+    {
+        ZoneOffsetTransition previousTransition = zoneRules.previousTransition(Instant.ofEpochMilli(instant));
+        if (previousTransition == null) {
+            // this is before the first transition
+            return instant;
+        }
+        return previousTransition.toEpochSecond() * 1000;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        JdkBasedDateTimeZone that = (JdkBasedDateTimeZone) o;
+        return Objects.equals(zoneRules, that.zoneRules);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(super.hashCode(), zoneRules);
```

Comment:
is hashcode compatible with equals? hashcode consults super's fields and equals does not (it would be correct for equals to do this)

Reply:
There is no field in the super class. Making things even more interesting, super does have an implementation for `hashCode`, but not `equals`. This is very uncommon. 

The `equals` and `hashCode` implementation here are generated by IntelliJ.

I think the implementation is correct as is. However, it indeed looks very suspicious.

Reply:
The super class declared `equals` as abstract so it must be implemented.

Reply:
Lets make this just `return getID().hashCode();`